### DISCUSSION TITLE
Simplify `infrastructure/`

### DISCUSSION THREAD

**Discussion by psschwei** [Qiskit/qiskit-serverless#625] [Ideas]
This is just some brainstorming around making the repo a little easier to maintain, specifically the `infrastructure/` directory. 

Right now `infrastructure/` has three subdirectories:
* [docker/](https://github.com/Qiskit-Extensions/quantum-serverless/tree/main/infrastructure/docker)
* [helm/](https://github.com/Qiskit-Extensions/quantum-serverless/tree/main/infrastructure/helm/quantumserverless)
* [terraform/](https://github.com/Qiskit-Extensions/quantum-serverless/tree/main/infrastructure/terraform)

I have suggestions for each one :smile: 

## Docker

Rather than maintaining a separate directory for the Dockerfiles, we should consider moving them with their components (i.e. the repository Dockerfile should live in `repository/`). This will also let us go back to naming them `Dockerfile`, which will let us drop the `-f` flag in the build command (and the build directory too). 

I'd also recommend dropping the `Makefile`, as I think the toil of maintaining it outweighs the benefit of a shorthand for `docker build/push` for local dev (we're not using it in the CI so should be fine to drop it).

## Helm

We can move this to a top-level `helm-charts/` directory, which will make it easier for users to find and shorten the URL to use when installing the released helm chart and make it a little easier to run the new install scripts.

## Terraform

Hot take: I would drop this completely. A team experienced enough to use terraform to set up a cluster will already have their own scripts to do so, and someone brand new to Kubernetes would in my opinion be better served by using the GUI of their chosen cloud provider to create a cluster.  I think it's totally fine for us to work from the assumption that users will already have a cluster when they come to us.

As an added plus, then we don't have to keep abreast of all the breaking terraform provider changes that seem to happen every couple of months... :wink: 

## 

Overall goal here is just to reduce the amount of things we need to maintain / update which should help reduce the toil for all of us (and give us more time to do all the other cool stuff we want to do). Figured a discussion was the appropriate place to discuss this :laughing: so curious on your thoughts @pacomf @IceKhan13 @Tansito @akihikokuroda

**Comment by pacomf** [comment]
I agree, but I would like to have the terraforms scripts in some folder in the repo at least during some months more... because we can help users with some knowledge in deployment, and it can be useful like templates and guides.

**Reply by psschwei** [reply]
Maybe we contribute them upstream to the IBM Cloud docs? 
https://cloud.ibm.com/docs/ibm-cloud-provider-for-terraform?topic=ibm-cloud-provider-for-terraform-tutorial-tf-clusters

**Reply by Tansito** [reply]
In relation with [my comment](https://github.com/Qiskit-Extensions/quantum-serverless/discussions/625#discussioncomment-6062013) I agree. Maybe once time we are more stable from the `terraform` part with `staging` & `production` we can move to that approach (open to discuss anyway).

**Comment by Tansito** [comment]
I like the idea @psschwei , thank you for the discussion !

Some thoughts from my part:

- **Docker**: totally agree, I think it has quite more sense that structure that the current one
- **Helm**: I agree too, at least in the meantime we can not dedicate a dedicated repo to it.
- **Terraform**: I almost agree with this. I think in a very near future we are not going to need it but for me right now is being useful to deploy the project when we have a demonstration or just for punctually testing purposes, for example.

My proposal with `terraform` is to wait until we have `staging` & `production` ready (around September) and decide there what to do, mostly because will be dev-ops in charge of this and not me. I don't know if it can be useful for testing though.

**Comment by IceKhan13** [comment]
Agree with everything. Even with terraform hot-take :)

**Comment by psschwei** [comment]
We did this in #868
